### PR TITLE
Introduce KOKKOS_PRINTF

### PR DESCRIPTION
**Pull request by masterleinad**
`printf` can't be used in SYCL kernels but needs an experimental feature that requires a slightly different syntax. This pull request aims to provide a version that can be used independent of the backend. It is up for discussion if this is something we want.
The current version doesn't work with an empty parameter list.

Based on top of #3609. https://github.com/kokkos/kokkos/commit/5eb7441deabf0a23a6ddea6fa34f65fd3e4908fe is the relevant commit.

### CODE REVIEW

**Comment by crtrott** [comment]
Just a quick reaction, but we should have a longer discussion: 

quick reaction: no, we shouldn't do this as a public feature, but rather have intel fix this. printf is the de-facto agreed upon way to get output from devices, and I think we can ask them to make this work. 

If we really want to support an abstraction for printing, I would argue we should introduce one which is truly independent of support for anything like printf. I.e. we give users a configurable user space class, which they have to capture and can print too, and then get the output on the host. With options to set maximum buffer size, control which threads actually print etc. Likely it would be more like std::ostream, i.e. you pipe stuff into it instead of having a mask like in printf and args. (The piping thing is I believe easier and cleaner to implement).

File: core/src/Kokkos_Macros.hpp [hunk 1]
```diff
@@ -382,6 +382,10 @@
 #define KOKKOS_IMPL_DEVICE_FUNCTION
 #endif
 
+#if !defined(KOKKOS_PRINTF)
+#define KOKKOS_PRINTF(format, ...) printf((format), __VA_ARGS__)
```
Comment (dhollman) [inline core/src/Kokkos_Macros.hpp:386]: The issue with variadic macros is _always_ going to be MSVC. I can never remember off of the top of my head what the differences are, but keep MSVC in mind before you go too far with it.

File: core/src/SYCL/Kokkos_SYCL.cpp [hunk 2]
```diff
@@ -93,7 +93,7 @@ bool SYCL::impl_is_initialized() {
 
 void SYCL::impl_finalize() { Impl::SYCLInternal::singleton().finalize(); }
 
-void SYCL::fence() const { m_space_instance->m_queue->wait(); }
+void SYCL::fence() const { m_space_instance->m_queue->wait_and_throw(); }
```
Comment (dhollman) [inline core/src/SYCL/Kokkos_SYCL.cpp:96]: Please elaborate on this change with a comment.

File: core/src/SYCL/Kokkos_SYCL_Instance.cpp [hunk 3]
```diff
@@ -96,7 +96,16 @@ void SYCLInternal::initialize(const cl::sycl::device& d) {
...
   const bool ok_dev  = true;
   if (ok_init && ok_dev) {
-    m_queue = std::make_unique<cl::sycl::queue>(d);
+    auto exception_handler = [](sycl::exception_list exceptions) {
+      for (std::exception_ptr const& e : exceptions) {
+        try {
+          std::rethrow_exception(e);
+        } catch (sycl::exception const& e) {
+          Kokkos::abort(e.what());
+        }
```
Comment (dhollman) [inline core/src/SYCL/Kokkos_SYCL_Instance.cpp:105]: I don't understand this completely. Are all of the exceptions in a `sycl::exception_list` derived from `sycl::exception`, or does the behavior of the handler depend on the order of the exceptions and throw out of itself sometimes?

File: core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp [hunk 4]
```diff
@@ -71,7 +71,7 @@ class Kokkos::Impl::ParallelFor<FunctorType, ExecPolicy,
         *space.impl_internal_space_instance();
     cl::sycl::queue& q = *instance.m_queue;
 
-    q.wait();
+    q.wait_and_throw();
```
Comment (dhollman) [inline core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp:74]: (Here and below) couldn't/shouldn't this just be `space.fence()`?

File: core/src/SYCL/Kokkos_SYCL.cpp (hunk 2 above)
Comment (masterleinad) [reply core/src/SYCL/Kokkos_SYCL.cpp:96]: I am happy to address these comments in #3609. :slightly_smiling_face:

File: core/src/Kokkos_Macros.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/Kokkos_Macros.hpp:386]: I think the (main) issue is described in https://stackoverflow.com/questions/5134523/msvc-doesnt-expand-va-args-correctly. It seems that there are workarounds to be able to use a variadic macro from another variadic macro and also using this macro is only necessary if also the SYCL backend should be supported.

**Review by nliber — APPROVED** [review]

**Comment by mhoemmen** [comment]
@crtrott wrote:
> If we really want to support an abstraction for printing, I would argue we should introduce one which is truly independent of support for anything like printf. I.e. we give users a configurable user space class, which they have to capture and can print too, and then get the output on the host. With options to set maximum buffer size....

Like `std::format_to_n`?  (Alas, it's not `nothrow`; it can throw `std::format_error`.)

**Comment by masterleinad** [comment]
I would still prefer having this as a function to be used internally until we have better functionality provided by SYCL/Intel.

**Comment by masterleinad** [comment]
Ready from my side. Needs at least one more review.

Again, this makes developing SYCL simpler and the macro should not be used outside of Kokkos.

File: core/src/Kokkos_Macros.hpp [hunk 5]
```diff
@@ -382,6 +382,10 @@
 #define KOKKOS_IMPL_DEVICE_FUNCTION
 #endif
 
+#if !defined(KOKKOS_IMPL_PRINTF)
+#define KOKKOS_IMPL_PRINTF(...) printf(__VA_ARGS__)
```
Comment (Rombur) [inline core/src/Kokkos_Macros.hpp:386]: I would add a comment that this macro is temporary and will disappear as soon as SYCL supports `printf` in kernels.

**Review by Rombur — APPROVED** [review]
I think this is fine as a temporary solution. If we know that Intel will support `printf` in the future, then I think it's OK to merge this. If Intel doesn't have any plan to support `printf` then we should think about a long term solution.

**Comment by masterleinad** [comment]
> I think this is fine as a temporary solution. If we know that Intel will support `printf` in the future, then I think it's OK to merge this. If Intel doesn't have any plan to support `printf` then we should think about a long term solution.

For me, the long term solution is a separate issue. It's a useful solution  (that might disappear at any point if/when we have found something better) to be used exclusively inside the library.

**Review by DavidPoliakoff — APPROVED** [review]
Approving on the condition that we continue to be annoying to people at Intel until we get `printf`

**Comment by masterleinad** [comment]
Retest this please.

File: containers/src/Kokkos_UnorderedMap.hpp [hunk 6]
```diff
@@ -540,10 +540,7 @@ class UnorderedMap {
           // Previously claimed an unused entry that was not inserted.
           // Release this unused entry immediately.
           if (!m_available_indexes.reset(new_index)) {
-            // FIXME_SYCL SYCL doesn't allow printf in kernels
-#ifndef KOKKOS_ENABLE_SYCL
-            printf("Unable to free existing\n");
-#endif
+            KOKKOS_IMPL_PRINTF("Unable to free existing\n");
```
Comment (dalg24) [inline containers/src/Kokkos_UnorderedMap.hpp:543]: Rename to KOKKOS_IMPL_PRINTF -> KOKKOS_IMPL_DO_NOT_USE_PRINTF

**Pull request merged**